### PR TITLE
Adding iree_atomic_slist_t type.

### PR DESCRIPTION
**Pull request by benvanik**
Well, a _somewhat atomic_ slist anyway :)
What matters today is that this _can_ be golfed down to single atomic ops, but this is blocking progress on threading work so I am playing it safe here.

### CODE REVIEW

File: iree/base/atomic_slist.c [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/base/atomic_slist.h"
+
+#include <assert.h>
+
+// TODO(benvanik): add TSAN annotations when switched to atomics:
+// https://github.com/gcc-mirror/gcc/blob/master/libsanitizer/include/sanitizer/tsan_interface_atomic.h
+// https://reviews.llvm.org/file/data/wade34tm2tsb66n2h26j/PHID-FILE-ogubmj4ak5fhbbtpb4go/D18500.id51829.diff
```
Comment (ScottTodd) [inline iree/base/atomic_slist.c:21]: 404? https://reviews.llvm.org/D18500 works
```suggestion
// https://reviews.llvm.org/D18500
```

File: iree/base/atomic_slist.h [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+// Typed and named wrappers for making atomic slists easier to work with.
+//
+// Usage:
+//  typedef struct {
+//    int some_fields;
+//    iree_atomic_slist_intrusive_ptr_t slist_next;
+//    int more_fields;
+//  } my_type_t;
+//  IREE_TYPED_ATOMIC_SLIST_WRAPPER(my_type, my_type_t,
+//                                  offsetof(my_type_t, slist_next));
```
Comment (ScottTodd) [inline iree/base/atomic_slist.h:189]: Neat, these macros work basically like a C++ templated class but in C?

**Review by ScottTodd — APPROVED** [review]
Wait until `benvanik-hack-and-slash` is merged, then rebase onto `main`?

File: iree/base/atomic_slist.h (hunk 2 above)
Comment (benvanik) [reply iree/base/atomic_slist.h:189]: yep! so you'll get nice type errors if you try to use the wrong types together and such.

**Pull request merged**